### ISSUE TITLE
[BUG] Wrong startIndex and endIndex rangeChanged callback in GroupedVirtuoso component

### ISSUE THREAD

**Issue by thiagoissao** [petyosi/react-virtuoso#712]
**Describe the bug**
Using `GroupedVirtuoso` with `rangeChanged` callback, the range `startIndex` and `endIndex` returned by the function sometimes emit wrong values. 

i have this summarized implementation: 
```ts
<GroupedVirtuoso
          data={users}
          style={{ height: 200 }}
          groupCounts={[500, 500]}
          rangeChanged={(range) => console.log(range)}
          ref={virtuoso}
          itemContent={(index, _, user) => {
            return (
              <div {...item} />
            );
          }}
        />
``` 

**Reproduction**
https://codesandbox.io/s/sandpack-project-forked-i1hkpv?file=/App.js

**To Reproduce**
1. Click on 'Scroll to 500'
2. See the log of the rangeChanged callback
3. If everything is normal, then scroll the list up and down

**Expected behavior**
I think the expected behavior is the rangeChanged emit only indexes that the list is rendering on the screen. And the `startIndex` cannot be greater than `endIndex`.
For example, in the image below:
- First log: The `startIndex: 497` and `endIndex: 1` should never happens because it `endIndex` cannot be smaller than `startIndex`
- Last log: The `startIndex: 1` and `endIndex: 502` should not be printed because all of this items list is not being rendered at this time. 

**Screenshots**
![image](https://user-images.githubusercontent.com/42099621/178733227-53d5377a-8854-4663-9ef5-b183ff935fd2.png)

**Desktop (please complete the following information):**
 - OS: Ubuntu Linux
 - Browser: Chrome

**Comment by petyosi** [comment]
Confirming that, it's a weird bug. I will look into it.

**Comment by thiagoissao** [comment]
Hi, good morning @petyosi!
Did you find the origin of this bug? Do you need any help?

**Comment by petyosi** [comment]
Things are quite busy on my side, so I'm unlikely to get to that in the next couple of days. I am fairly certain that this might be related to the group items having `index` which is the index of the groups, but I am not sure. Happy to accept a PR.

**Comment by thiagoissao** [comment]
Hi @petyosi, i'm looking the source code and i probably found the bug. I tested it locally and the problem really was what you said.
In the `listStateSystem.ts` we have a function called `rangeChanged`. Inside this function, we have `startIndex` and `endIndex`  that sometimes return the groupIndex value. 

I fix this bug in a "brute way" and it looks like ok, but only for this specific bug. 

The code below solves the bug, but it has two problems:
- Many automated tests have broken
- The performance of the new implementation probably is not ok , because inside map function i inserted these "for" loops: `find`, `reverse` and `find` again 

```ts
    const rangeChanged = u.streamFromEmitter(
      u.pipe(
        listState,
        u.filter(({ items }) => items.length > 0),
        u.map(({ items }) => {
          const startIndex = items.find((item) => item.type !== 'group')
          const endIndex = items.reverse().find((item) => item.type !== 'group')
          return {
            startIndex: startIndex?.index || items[0].index,
            endIndex: endIndex?.index || items[items.length - 1].index,
          } as ListRange
        }),
        u.distinctUntilChanged(rangeComparator)
      )
    )
```

So, what do you recommend in this case?

**Comment by petyosi** [comment]
Good news. The find/reverse optimized with a `while item.type !== 'group'` + index increment/decrement. Strange that the tests are broken, most likely due to the `items.reverse()` modifying the array in place.

**Comment by thiagoissao** [comment]
This is the new code:
```ts
const rangeChanged = u.streamFromEmitter(
      u.pipe(
        listState,
        u.filter(({ items }) => items.length > 0),
        u.map(({ items }) => {
          let startIndex = 0
          while (items[startIndex].type === 'group') startIndex++

          let endIndex = items.length - 1
          while (items[endIndex].type === 'group') endIndex--

          return {
            startIndex: items[startIndex].index,
            endIndex: items[endIndex].index,
          } as ListRange
        }),
        u.distinctUntilChanged(rangeComparator)
      )
    )
```
it works as expected :clap:  

But, i couldn't push it in a new fix branch, how can i generate a new PR?

**Comment by petyosi** [comment]
Great. Fork the repo, push your branch there, and GH should suggest the way: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

**Comment by petyosi** [comment]
Fixed in #713
